src/adera/Machines/RCSController.cpp

```diff
@@ -0,0 +1,125 @@
+#include "RCSController.h"
+#include "osp/Active/ActiveScene.h"
+
+#include <Magnum/Math/Vector.h>
+
+using namespace adera::active::machines;
+using namespace osp::active;
+using Magnum::Vector3;
+
+
+MachineRCSController::MachineRCSController()
+    : Machine(true),
+    m_wiCommandOrient(this, "Orient"),
+    m_woThrottle(this, "Throttle")
+{
+    using wiretype::Percent;
+    m_woThrottle.value() = Percent{0.0f};
+}
+
+void MachineRCSController::propagate_output(WireOutput* output)
+{
+
+}
+
+WireInput* MachineRCSController::request_input(osp::WireInPort port)
+{
+    return existing_inputs()[port];
+}
+
+WireOutput* MachineRCSController::request_output(osp::WireOutPort port)
+{
+    return existing_outputs()[port];
+}
+
+std::vector<WireInput*> MachineRCSController::existing_inputs()
+{
+    return {&m_wiCommandOrient};
+}
+
+std::vector<WireOutput*> MachineRCSController::existing_outputs()
+{
+    return {&m_woThrottle};
+}
+
+SysMachineRCSController::SysMachineRCSController(ActiveScene& scene)
+    :SysMachine<SysMachineRCSController, MachineRCSController>(scene),
+    m_updateControls(scene.get_update_order(), "mach_rcs", "wire", "controls",
+        [this](ActiveScene& rScene) { this->update_controls(rScene); })
+{
+}
+
+float SysMachineRCSController::thruster_influence(Vector3 posOset, Vector3 direction,
+    Vector3 cmdTransl, Vector3 cmdRot)
+{
+    using Magnum::Math::cross;
+    using Magnum::Math::dot;
+    using Magnum::Math::clamp;
+
+    Vector3 thrust = -direction.normalized();
+    
+    float rotInfluence = 0.0f;
+    float translInfluence = 0.0f;
+
+    if (cmdRot.length() > 0.0f)
+    {
+        Vector3 torque = cross(posOset, thrust).normalized();
+        rotInfluence = dot(torque, cmdRot.normalized());
+    }
+
+    if (cmdTransl.length() > 0.0f)
+    {
+        translInfluence = dot(thrust, cmdTransl.normalized());
+    }
+
+    float total = rotInfluence + translInfluence;
+
+    if (total < .01f)
+    {
+        // Ignore small contributions from imprecision
+        // Real thrusters can't throttle this deep anyways
+        return 0.0f;
+    }
+    return clamp(rotInfluence + translInfluence, 0.0f, 1.0f);
+}
+
+void SysMachineRCSController::update_controls(ActiveScene& rScene)
+{
+    auto view = m_scene.get_registry().template view<MachineRCSController, ACompTransform>();
```

Comment:
I know this is a thing...

But why have I gone almost my.whole career only using the template keyword to disambiguate a function call and OSP uses it everywhere?

Reply:
I'm not sure why there's a template keyword here. It's not ambiguous and it compiles just fine without it.

Reply:
I only add it when it doesn't compile... I can try removing it again though.

Reply:
You're using VS2019 right? Double check you're up to date?

Reply:
I removed it and it compiles fine, but the problem is not my local compilation; MSVC will compile pretty egregiously noncompliant C++, the issue is that when I push the code the build fails on Clang or gcc or something. I'll let you know what happens when the build runs next time.